mathjax3-ts/output/chtml/FontData.ts

```diff
@@ -0,0 +1,328 @@
+     *   (The idea is to be able to have something like bold-italic inherit from both
+     *   bold and intalic by having the prototype chain include a copy of bold plus
+     *   the full italic chain.  So if something is not defined explicitly in bold-italic,
+     *   it defaults first to a bold version, than an italic version, then the normal
+     *   version, which is in the italic prototype chain.)
+     *
+     * @param{string} name     The new variant to create
+     * @param{string} inherit  The variant to use if a character is not in this one
+     * @param{string} link     A variant to search before the inherit one (but only
+     *                           its top-level object).
+     */
+    public createVariant(name: string, inherit: string = null, link: string = null) {
+        let variant = {
+            linked: [] as CharMap[],
+            chars: (inherit ? Object.create(this.variant[inherit].chars) : {}) as CharMap
+        };
+        if (link && this.variant[link]) {
+            Object.assign(variant.chars, this.variant[link].chars);
+            this.variant[link].linked.push(variant.chars);
+            variant.chars = Object.create(variant.chars);
+        }
+        this.variant[name] = variant;
+    }
+
+    /*
+     * Create a collection of variants
+     *
+     * @param{string[][]} variants  Array of [name, inherit?, link?] values for
+     *                              the variants to define
+     */
+    public createVariants(variants: string[][]) {
+        for (const variant of variants) {
+            this.createVariant.apply(this, variant);
+        }
+    }
+
+    /*
+     * Defines new character data in a given variant
+     *
+     * @param{srtring} name   The variant for these characters
+     * @param{CharMap} chars  The characters to define
+     */
+    public defineChars(name: string, chars: CharMap) {
+        let variant = this.variant[name];
+        Object.assign(variant.chars, chars);
+        for (const link of variant.linked) {
+            Object.assign(link, chars);
+        }
+    }
+
+    /*
+     * Defines strety delimiters
+     *
+     * @param{DelimiterMap} delims  The delimiters to define
+     */
+    public defineDelimiters(delims: DelimiterMap) {
+        Object.assign(this.delimiters, delims);
+    }
+
+    /*
+     * @param{number} n  The delimiter character number whose data is desired
+     * @return{DelimiterData}  The data for that delimiter (or null)
+     */
+    public getDelimiter(n: number) {
+        return this.delimiters[n];
+    }
+
+    /*
+     * @param{number} n  The delimiter character number whose variant is needed
+     * @param{number} i  The index in the size array of the size whose variant is needed
+     * @return{string}   The variant of the i-th size for delimiter n
+     */
+    public getSizeVariant(n: number, i: number) {
+        if (this.delimiters[n].variants) {
+            i = this.delimiters[n].variants[i];
```

Comment:
This can be undefined. Is that a problem?

Reply:
Note that `getSizeVariant()` is only called when the character is known to be stretchy, i.e., `this.delimiters[n]` is known to exist, so isn't undefined.  For a properly constructed `DelimiterData`, the `variants` array should be the same length as the `sizes` array, but you are correct that there is no guarantee of that.

One could, I suppose, use
```
i = this.delimiters[n].variants[i] || 0;
```
but I think I'd rather get an error when the `undefined` is used as a number in order to find out that the data are not correct rather than hiding it and getting a wrong glyph.